### ISSUE TITLE
Image generation fails with WP Offload Media

### ISSUE THREAD

**Issue by scoutbranding** [Booreiland/responsive-pics#5]
**Describe the bug**
Image generation fails when using [Delicious Brains WP Offload Media](https://deliciousbrains.com/wp-offload-media/) and Amazon S3.

**To Reproduce**
1. Activate WP Offload Media using Amazon S3.
2. Call `ResponsivePics::get()`.
3. See error:

```
ErrorException thrown with message "chmod(): DeliciousBrains\WP_Offload_Media\Providers\Streams\AWS_S3_Stream_Wrapper::stream_metadata is not implemented! (View: /srv/www/example.com/current/web/app/themes/example/resources/views/components/image.blade.php) (View: /srv/www/example.com/current/web/app/themes/example/resources/views/components/image.blade.php) (View: /srv/www/example.com/current/web/app/themes/example/resources/views/components/image.blade.php)"

Stacktrace:
#30 ErrorException in /srv/www/example.com/current/web/wp/wp-includes/class-wp-image-editor-gd.php:447
#29 ErrorException in /srv/www/example.com/current/web/wp/wp-includes/class-wp-image-editor-gd.php:447
#28 ErrorException in /srv/www/example.com/current/web/wp/wp-includes/class-wp-image-editor-gd.php:447
#27 ErrorException in /srv/www/example.com/current/web/wp/wp-includes/class-wp-image-editor-gd.php:447
#26 chmod in /srv/www/example.com/current/web/wp/wp-includes/class-wp-image-editor-gd.php:447
#25 WP_Image_Editor_GD:_save in /srv/www/example.com/current/web/wp/wp-includes/class-wp-image-editor-gd.php:400
#24 WP_Image_Editor_GD:save in /srv/www/example.com/current/web/app/themes/example/vendor/booreiland/responsive-pics/ResponsivePics.php:472
#23 ResponsivePics:get_resized_url in /srv/www/example.com/current/web/app/themes/example/vendor/booreiland/responsive-pics/ResponsivePics.php:575
#22 ResponsivePics:get_definition in /srv/www/example.com/current/web/app/themes/example/vendor/booreiland/responsive-pics/ResponsivePics.php:832
#21 ResponsivePics:get in /srv/www/example.com/current/web/app/themes/example/storage/framework/views/b1f0872d7405600aa7373fb2cc9653e0bb6645bf.php:12
#20 include in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/Engines/PhpEngine.php:43
#19 Illuminate\View\Engines\PhpEngine:evaluatePath in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/Engines/CompilerEngine.php:59
#18 Illuminate\View\Engines\CompilerEngine:get in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:143
#17 Illuminate\View\View:getContents in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:126
#16 Illuminate\View\View:renderContents in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:91
#15 Illuminate\View\View:render in /srv/www/example.com/current/web/app/themes/example/storage/framework/views/10904c72cb6518c89c60227fe34aa6a028995874.php:9
#14 include in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/Engines/PhpEngine.php:43
#13 Illuminate\View\Engines\PhpEngine:evaluatePath in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/Engines/CompilerEngine.php:59
#12 Illuminate\View\Engines\CompilerEngine:get in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:143
#11 Illuminate\View\View:getContents in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:126
#10 Illuminate\View\View:renderContents in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:91
#9 Illuminate\View\View:render in /srv/www/example.com/current/web/app/themes/example/storage/framework/views/74de89f46a5e7c967db7d52de3fa06d066dfef48.php:4
#8 include in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/Engines/PhpEngine.php:43
#7 Illuminate\View\Engines\PhpEngine:evaluatePath in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/Engines/CompilerEngine.php:59
#6 Illuminate\View\Engines\CompilerEngine:get in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:143
#5 Illuminate\View\View:getContents in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:126
#4 Illuminate\View\View:renderContents in /srv/www/example.com/current/web/app/themes/example/vendor/illuminate/view/View.php:91
#3 Illuminate\View\View:render in /srv/www/example.com/current/web/app/themes/example/index.php:10
#2 include in /srv/www/example.com/current/web/wp/wp-includes/template-loader.php:98
#1 require_once in /srv/www/example.com/current/web/wp/wp-blog-header.php:19
#0 require in /srv/www/example.com/current/web/index.php:6
```

**Comment by scoutbranding** [comment]
I understand this might not be within scope of this plugin, but it would be amazing if it could work copesceptically with external CDNs like S3.

**Comment by Twansparant** [comment]
> Image generation fails when using Delicious Brains WP Offload Media and Amazon S3.
This makes sense since it fails when it tries to save the image, which exists on an external storage and not in the wordpress uploads.

> I understand this might not be within scope of this plugin
Unfortunately this is out of scope indeed. This plugin was build purely for easier resizing and cropping images within the wordpress environment and storage. For resizing images in an external storage, you could use something like (Cloudinary)[https://cloudinary.com/]

**Comment by Twansparant** [comment]
Hi @scoutbranding,

We're currently in the need for this feature as well, so I created a `dev-feat/s3-offload` branch that adds support for the [WP Offload Media Lite](https://nl.wordpress.org/plugins/amazon-s3-and-cloudfront/) plugin. It seems to be working fine, but I was only able to test this with the free **Lite** version of the plugin.

Would you be willing to test our `dev-feat/s3-offload` branch and see if you come up with any issues? And if you're using the **Pro** version, could you report any errors/issues that you bump into?

You can use the branch like this:
```
"booreiland/responsive-pics": "dev-feat/s3-offload"
```
Many thanks in advance!

**Comment by joshuafredrickson** [comment]
Since this issue was posted, we've worked around it by using Imgix. However, I can set up a test env and see how it works with the Pro version.

**Comment by Twansparant** [comment]
Hi @joshuafredrickson, apologies for my late reply;
Are you still willing to try out our `feat/s3-offload` branch in a test environment?
Would help us a lot! Thanks

**Comment by joshuafredrickson** [comment]
Unfortunately, it looks like the pro version doesn't use `upload_attachment()`:

```
Fatal error: Uncaught Error: Call to undefined method Amazon_S3_And_CloudFront_Pro::upload_attachment() in /srv/www/site/current/web/app/plugins/responsive-pics/src/class-s3-offload.php:11
```

There is an `integrations` folder in the pro plugin folder, which may provide some inspiration and might be worth exploring.

**Comment by Twansparant** [comment]
Ok, that error you got is not due to the **Pro** version, but because of their plugin update from **2.5.5** to **2.6+**.
We used their function `upload_attachment` but that doesn't exist anymore in version **2.6+**.
I'm currently looking for the equivalent function for the versions 2.6+.

**Comment by joshuafredrickson** [comment]
That makes sense! Hopefully there's a bit more parity between their free and pro versions. 👍

**Comment by Twansparant** [comment]
Fixed in [version 1.6.0](https://github.com/Booreiland/responsive-pics/releases/tag/1.6.0)!
